distribution/pull_v2.go

```diff
@@ -877,6 +881,30 @@ func (p *v2Puller) pullSchema2Config(ctx context.Context, dgst digest.Digest) (c
 	return configJSON, nil
 }
 
+func retry(ctx context.Context, attempts int, sleep time.Duration, f func(ctx context.Context) error) (err error) {
+	for i := 0; i < attempts; i++ {
+		err = retryOnError(f(ctx))
+		if err == nil {
+			return nil
+		}
+		if _, isDNR := err.(xfer.DoNotRetry); isDNR {
+			logrus.WithError(err).Errorf("download failed after %d attempts", attempts)
+			return err
```

Comment:
Actually; instead of returning the error, would it make sense to `break` ? (which would return the error as well - at the end of the function)

Reply:
I'm not sure I follow the last suggestion. I've named the variables in the retry loop better, however, and made the logging as requested.

Reply:
So currently, the code is something like:

```go
for attempt := 0; attempt < maxAttempts; attempt++ {
	// if no error return nil
	// if no-retry error, return the error
}

return errors.Wrapf(err, "download failed after attempts=%d", maxAttempts)
```

So, the error is only "decorated" if we failed after `maxAttempt`; my thinking was that it could be useful to include the number of attempts also if we fail before that, so something like:

```go
attempt := 0
for ; attempt < maxAttempts; attempt++ {
	// if no error return nil
	if xfer.IsDoNotRetryError(err) {
		break
	}
}

return errors.Wrapf(err, "download failed after %d attempts", attempt+1)
```

Reply:
Done!